Detectors/CTP/Raw/test/CTPtestRawWriter.cxx

```diff
@@ -0,0 +1,173 @@
+    for (const auto& ir : irs) {
+      int nCRU2Fill = writer.isCRUDetector() ? NCRU - 1 : NCRU; // in CRU mode we will fill 1 special CRU with preformatted data
+      for (int icru = 0; icru < nCRU2Fill; icru++) {
+        // we will create non-0 payload for all but 1st link of every CRU, the writer should take care
+        // of creating empty HBFs for the links w/o data
+        for (int il = 0; il < NLinkPerCRU; il++) {
+          buffer.clear();
+          //int nGBT = gRandom->Poisson(RDHUtils::MAXCRUPage / RDHUtils::GBTWord * (il));
+          if (nGBT) {
+            buffer.resize((nGBT + 2) * RDHUtils::GBTWord, icru * NLinkPerCRU + il); // reserve 16B words accounting for the Header and Trailer
+            std::memcpy(buffer.data(), PLHeader.c_str(), RDHUtils::GBTWord);
+            std::memcpy(buffer.data() + buffer.size() - RDHUtils::GBTWord, PLTrailer.c_str(), RDHUtils::GBTWord);
+            // we don't care here about the content of the payload, except the presence of header and trailer
+          }
+          writer.addData((icru << feeIDShift) + il, icru, il, 0, ir, buffer);
+        }
+      }
+    }
+    if (writer.isCRUDetector()) {
+      // fill special CRU with preformatted pages
+      auto irHB = HBFUtils::Instance().getFirstIR(); // IR of the TF0/HBF0
+      int cruID = NCRU - 1;
+      while (irHB < irs.back()) {
+        for (int il = 0; il < NLinkPerCRU; il++) {
+          buffer.clear();
+          int pgSize = SpecSize[il] - sizeof(RDHAny);
+          buffer.resize(pgSize);
+          for (int ipg = 2 * (NLinkPerCRU - il); ipg--;) {                       // just to enforce writing multiple pages per selected HBFs
+            writer.addData((cruID << 8) + il, cruID, il, 0, irHB, buffer, true); // last argument is there to enforce a special "preformatted" mode
+            nPreformatPages++;
+          }
+        }
+        irHB.orbit += HBFUtils::Instance().getNOrbitsPerTF() / NPreformHBFPerTF; // we will write 32 such HBFs per TF
+      }
+    }
+
+    // for further use we write the configuration file
+    writer.writeConfFile(writer.getOrigin().str, "RAWDATA", configName);
+    writer.close(); // flush buffers and close outputs
+  }
+  // optional callback functions to register in the RawFileWriter
+  //_________________________________________________________________
+  void emptyHBFMethod(const RDHAny* rdh, std::vector<char>& toAdd) const
+  {
+    // what we want to add for every empty page
+    toAdd.resize(RDHUtils::GBTWord);
+    std::memcpy(toAdd.data(), HBFEmpty.c_str(), RDHUtils::GBTWord);
+  }
+
+  //_________________________________________________________________
+  int carryOverMethod(const RDHAny* rdh, const gsl::span<char> data, const char* ptr, int maxSize, int splitID,
+                      std::vector<char>& trailer, std::vector<char>& header) const
+  {
+    // how we want to split the large payloads. The data is the full payload which was sent for writing and
```

Comment:
do you really split the data for CTP?

Reply:
No, as comment before. CTPtestRawWriter.cxx should not have been commited. Sorry.

Reply:
IN summary,  CTPtestRawWriter.cxx  is there by mistake. Macro writeReadCTPRaw.C with classes is supposed to do similar job.